### ISSUE TITLE
Allow rebasing onto any ref

### ISSUE THREAD

**Issue by jesseduffield** [jesseduffield/lazygit#2850]
**Is your feature request related to a problem? Please describe.**
Currently the 'r' keybinding only brings up the rebasing menu in the branches view and remote branches view.

**Describe the solution you'd like**
I want it to bring up the rebasing menu in any view that contains something you can rebase onto which means:
* commits
* sub-commits
* reflog commits
* tags

I think that's everything.

The tricky part is that the 'r' keybinding is already used to support rewording a commit from within lazygit. 'R' is used for rewording a commit in an editor. We have other places where 'R' is used for renaming e.g. the branches view. So I propose that 'R' now opens a menu with an option for rewording from within lazygit ('i') and rewording from an external editor ('e').

Yes, nobody is more upset by extra keypresses than me, but I think this is necessary. I'm pretty sure that people will sometimes want the external editor and sometimes want the internal editor, so having a config for it wouldn't work, unless the config was a choice between three options: always-internal, always-external, and show-menu. But I don't want to do that unless people actually want it.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by stefanhaller** [comment]
The proposal makes sense to me in general.

I don't understand the section "Describe alternatives you've considered", it seems to describe the same option again?

As for shuffling keybindings around big time, I'd be wary of muscle memory; remapping something like `m` to `M` seems problematic for me for that reason.

And finally, as for having a menu behind `R`: here I'm undecided. The only reason why I'm not always using rename-with-editor now, but sometimes use the internal editor is that this saves a few milliseconds of time when I want to work really fast (because VS Code is annoyingly slow to come up). The extra menu would slow that down again, so I might as well have a config that always brings up the editor.

**Comment by jesseduffield** [comment]
> I don't understand the section "Describe alternatives you've considered", it seems to describe the same option again?

Whoops: in the main proposal I meant to say that 'R' should rebase and 'r' should reword, and the alternative is to swap them around.

**Comment by jesseduffield** [comment]
> And finally, as for having a menu behind R: here I'm undecided. The only reason why I'm not always using rename-with-editor now, but sometimes use the internal editor is that this saves a few milliseconds of time when I want to work really fast (because VS Code is annoyingly slow to come up). The extra menu would slow that down again, so I might as well have a config that always brings up the editor.

Alright let's add the config

**Comment by stefanhaller** [comment]
> > I don't understand the section "Describe alternatives you've considered", it seems to describe the same option again?
> 
> Whoops: in the main proposal I meant to say that 'R' should rebase and 'r' should reword, and the alternative is to swap them around.

I find this very tricky to decide. What's more important: consistency of keybindings across panels, or having the simplest keybindings for the most frequent commands in each panel? 

In the local commits panel, you will almost never use rebase, but rewording is a very frequent operation. Conversely, in the branches panel, rebasing is much more common than rewording (for me at least).

**Comment by jesseduffield** [comment]
I think consistency across panels is more important

**Comment by stefanhaller** [comment]
Yeah, probably. Just one more thought about muscle memory: swapping `r` and `R` can be dangerous for existing users, because both rebase and reword-with-editor have the `r enter` (or `R enter`) sequence. So for people who like to work fast and hit enter quickly without paying attention to the popup that comes up, this can easily do the wrong thing.

It's probably not enough reason to not swap them, and I also don't have a good suggestion what to do about it (except to tell people very loudly about the change in release notes). Just wanted to mention it.

**Comment by jesseduffield** [comment]
Agreed about being loud about the change (I'm thinking a popup upon opening lazygit advising them to check the release notes). I think it would be good to do an audit of other keybindings that could be made consistent so that we can make the change in one fell swoop. Machiavelli said if you're going to do good, space it out, but if you're going to cause harm, do it all at once.

**Comment by sedmicha** [comment]
Wow, what a coincidence. I was just about to open a feature request asking for this functionality, and it's already there, by the creator itself :)

I like the idea of having a config option to choose the default reword action instead of always having a menu too. 


Also, would it perhaps be feasible to add a shortcut to the built-in reword edit to switch you to an editor instead? 

I often find myself pressing 'r', realizing that I could really use an editor, so I 'Esc' 'R'. It would be nice if I could just press Ctrl+r or something instead and it'd switch me to the editor without having to cancel out. 

This isn't really needed now, as Esc R isn't that much of a hassle, but if there was a popup menu in-between, it's starting to be a bit cumbersome. This way you could also set the default action to the built-in reword edit and switch to the editor from there if needed. 

Just a thought. I could probably live with just reword-with-editor as the default, but this would be nice to have.

**Comment by stefanhaller** [comment]
> Also, would it perhaps be feasible to add a shortcut to the built-in reword edit to switch you to an editor instead?
> 
> I often find myself pressing 'r', realizing that I could really use an editor, so I 'Esc' 'R'. It would be nice if I could just press Ctrl+r or something instead and it'd switch me to the editor without having to cancel out.

@sedmicha I have something for you 😄.  See #2881.

**Comment by sedmicha** [comment]
> > Also, would it perhaps be feasible to add a shortcut to the built-in reword edit to switch you to an editor instead?
> > I often find myself pressing 'r', realizing that I could really use an editor, so I 'Esc' 'R'. It would be nice if I could just press Ctrl+r or something instead and it'd switch me to the editor without having to cancel out.
> 
> @sedmicha I have something for you 😄. See #2881.

Wow, thank you so much. I wouldn't have even expected a response this fast, let alone a PR already implementing it 😅. 

You guys are doing an amazing job here, truly.
